File: Christmas-Lights/c++/src/LightGrid.cpp
#include "LightGrid.h"

#include <iomanip>
#include <iostream>
#include <typeinfo> 

LightGrid::LightGrid(size_t maxX, size_t maxY)
    : m_maxX{maxX}, m_maxY{maxY}
{
}

size_t LightGrid::getCountOfLightsOn() const
{
    size_t countOfLightsOn{0};
    for (size_t x{0}; x <= m_maxX; ++x)
    {
        for (size_t y{0}; y <= m_maxY; ++y)
        {
            if (isOn(x, y))
            {
                ++countOfLightsOn;
            }
        }
    }

    return countOfLightsOn;
}

size_t LightGrid::getTotalBrightness() const
{
    size_t totalBrightness{0};
    for (size_t x = 0; x <= m_maxX; ++x)
    {
        for (size_t y = 0; y <= m_maxY; ++y)
        {
            totalBrightness += getBrightness(x, y);
        }
    }

    return totalBrightness;
}

const LightGrid *LightGrid::showSummary() const
{
    std::cout << typeid(*this).name()
              << " lights on: " << std::setw(6) << getCountOfLightsOn()
              << " | total brightness: " << std::setw(12) << getTotalBrightness()
              << std::endl;

    return this;
}

File: Christmas-Lights/php/src/LightGrid.php
<?php

abstract class LightGrid
{
    final public static function start(int $width, int $height): LightGrid
    {
        return new EmptyGrid($width, $height);
    }

    protected function __construct(
        protected readonly int $width,
        protected readonly int $height
    )
    {
    }

    public function __toString(): string
    {
        return sprintf('%s [%d, %d]', get_class($this), $this->width, $this->height);
    }


    abstract protected function getBrightness(int $x, int $y): int;

    abstract protected function isOn(int $x, int $y): bool;

    final public function showSummary(): void
    {
        echo "$this\n";
        $this->showCountLightsOn();
        echo ' | ';
        $this->showSumBrightness();
        echo "\n";
    }

    private function showCountLightsOn(): void
    {
        $count = 0;
        for ($x = 0; $x <= $this->width; ++$x) {
            for ($y = 0; $y <= $this->height; ++$y) {
                if ($this->isOn($x, $y)) {
                    ++$count;
                }
            }
        }

        printf("lights on: %7d", $count);
    }

    private function showSumBrightness(): void
    {
        $brightness = 0;
        for ($x = 0; $x <= $this->width; ++$x) {
            for ($y = 0; $y <= $this->height; ++$y) {
                $brightness += $this->getBrightness($x, $y);
            }
        }

        printf("total brightness: %12d", $brightness);
    }

    final public function toggle(Rectangle $selection): LightGrid
    {
        return new Toggle($this, $selection);
    }

    final public function turnOff(Rectangle $selection): LightGrid
    {
        return new TurnOff($this, $selection);
    }

    final public function turnOn(Rectangle $selection): LightGrid
    {
        return new TurnOn($this, $selection);
    }
}